src/main/java/de/metas/ui/web/handlingunits/process/WEBUIHUCreationWithSerialNumberService.java

```diff
@@ -0,0 +1,262 @@
+			}
+		}
+		else
+		{
+			Check.assume(parentRow.isTU(), " Parent row must be TU: " + parentRow);
+			I_M_HU parentHU = parentRow.getM_HU();
+
+			if (isAggregateHU(parentRow))
+			{
+				final HUEditorRow luRow = huEditorRowHierarchy.getTopLevelRow();
+
+				parentHU = createNonAggregatedTU(parentRow, luRow);
+
+				huToSplit = handlingUnitsDAO.retrieveIncludedHUs(parentHU).get(0);
+			}
+
+			final int tuCapacity = calculateTUCapacity(parentHU);
+			numberOfCUsToCreate = getMinimumOfThree(tuCapacity, maxCUsAllowedPerBatch, initialQtyCU);
+			for (int i = 0; i < numberOfCUsToCreate; i++)
+			{
+				final List<I_M_HU> createdCUs = newHUTransformation().cuToExistingTU(huToSplit, BigDecimal.ONE, parentHU);
+
+				splitCUIDs.addAll(createdCUs.stream().map(I_M_HU::getM_HU_ID).collect(ImmutableSet.toImmutableSet()));
+			}
+		}
+
+		return splitCUIDs;
+	}
+
+	private int calculateTUCapacity(final I_M_HU parentHU)
+	{
+		final I_M_HU_PI_Item_Product huPIP = parentHU.getM_HU_PI_Item_Product();
+
+		if (huPIP == null)
+		{
+			// TODO
+			throw new UnsupportedOperationException();
+		}
+
+		return huPIP.getQty().intValueExact();
+	}
+
+	private int getMinimumOfThree(final int no1, final int no2, final int no3)
+	{
+		return no1 < no2 ? (no1 < no3 ? no1 : no3) : (no2 < no3 ? no2 : no3);
+	}
+
+	private I_M_HU createNonAggregatedTU(final HUEditorRow tuRow, final HUEditorRow luRow)
+	{
+		final I_M_HU newTU = newHUTransformation().tuToNewTUs(tuRow.getM_HU(), BigDecimal.ONE).get(0);
+
+		if (luRow != null)
+		{
+			moveTUToLU(newTU.getM_HU_ID(), luRow, tuRow);
+		}
+
+		return newTU;
+	}
+
+	private void moveTUToLU(final int tuId, final HUEditorRow luRow, final HUEditorRow tuRow)
+	{
+		final I_M_HU newTU = create(Env.getCtx(), tuId, I_M_HU.class, ITrx.TRXNAME_ThreadInherited);
+		newHUTransformation().tuToExistingLU(newTU, BigDecimal.ONE, luRow.getM_HU());
+	}
+
+	private void assignSerialNumbersToCUs(final Set<Integer> cuIDs, final List<String> availableSerialNumbers)
+	{
+
+		final List<Integer> listOfCUIDs = cuIDs.stream().collect(ImmutableList.toImmutableList());
+
+		final int numberOfCUs = listOfCUIDs.size();
+
+		for (int i = 0; i < numberOfCUs; i++)
+		{
+			if (availableSerialNumbers.isEmpty())
+			{
+				return;
+			}
+			
+			assignSerialNumberToCU(listOfCUIDs.get(i), availableSerialNumbers.remove(0));
+		}
+	}
+
+	private void assignSerialNumberToCU(final int huId, final String serialNo)
+	{
+		final I_M_HU hu = create(Env.getCtx(), huId, I_M_HU.class, ITrx.TRXNAME_ThreadInherited);
+
+		final IContextAware ctxAware = getContextAware(hu);
+
+		final IHUContext huContext = handlingUnitsBL.createMutableHUContext(ctxAware);
+
+		final IAttributeStorage attributeStorage = getAttributeStorage(huContext, hu);
+
+		final I_M_Attribute serialNoAttribute = serialNoDAO.getSerialNoAttribute(ctxAware.getCtx());
+
+		Check.assume(attributeStorage.hasAttribute(serialNoAttribute), "There is no SerialNo attribute defined for the handling unit" + hu);
```

Comment:
* the mesasage of `assume` should state the assumption which is violated. If you want to state the error (i.e. the violation itself), please use `errorIf` or `errorUnless`
* please don't concatenate strings like this..instead, you could to `.., "There is no SerialNo attribute defined for the handling unit {}", hu);`.
* i recommend to also output the `serialNo` parameter, because you can, and it *might* help, if the problem occurs

Reply:
Used Check.errorUnless. String concatenation fixed